pages/quests/index.tsx

```diff
@@ -8,24 +8,111 @@ import {
-  const { keyword, inputChange, submit, isSubmitting } = useKeywordSubmit()
+  const { keyword, inputChange, submit, isSubmitting, keywordSubmitSucceeded } = useKeywordSubmit()
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [newTokenImageURI, setNewTokenImageURI] = useState('')
+  const [newTokenJSON, setNewTokenJSON] = useState<KamonToken>()
+  const [updateTxFailed, setUpdateTxFailed] = useState<boolean>()
+
+  const { ipfsSubmit, ipfsSubmitSucceeded, ipfsSubmitIsSubmitting } = useIpfsSubmit()
+  const { ipfsGet, ipfsGetIsSubmitting } = useIpfsGet()
+  const { updateToken, updateTokenSucceeded, updateTokenIsSubmitting } = useUpdateToken(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+  const submittingSomething = () => {
+    return isSubmitting || ipfsSubmitIsSubmitting || ipfsGetIsSubmitting || updateTokenIsSubmitting
+  }
+  useEffect(() => {
+    if (tokenURI) {
+      const fetchData = async () => {
+        const pinataRequest = await fetch(tokenURI.toString())
+        const responseJson = await pinataRequest.json()
+        setTokenJSON(responseJson)
+      }
+      fetchData()
+    }
+  }, [tokenURI])
+
+  useEffect(() => {
+    if (keywordSubmitSucceeded && !ipfsSubmitIsSubmitting
+      && ipfsSubmitSucceeded == undefined
+    ) {
+      if (data?.address == undefined || tokenJSON == undefined) return
+      const userAddress: string = data?.address
+      const ipfsSubmitWrapper = async () => {
+        const ipfsSubmitRet = await ipfsSubmit(tokenJSON, userAddress)
+        if (ipfsSubmitRet !== 'error' && ipfsSubmitRet !== 'same_points') {
+          const tempFinalTokenUri = await ipfsSubmitRet
+          setFinalTokenUri(tempFinalTokenUri)
+        }
+      }
+      setUpdateTxFailed(false)
+      ipfsSubmitWrapper()
+    }
+  }, [keywordSubmitSucceeded, data?.address, tokenJSON, ipfsSubmit, ipfsSubmitIsSubmitting, ipfsSubmitSucceeded])
+
+  useEffect(() => {
+    if (finalTokenUri && !ipfsGetIsSubmitting && newTokenJSON == undefined) {
+      const getNewToken = async () => {
+        const ipfsGetRet = await ipfsGet(finalTokenUri)
+        if (ipfsGetRet == 'error') {
+          return
+        }
+        setNewTokenJSON(ipfsGetRet)
+        const theTokenId = tokenIdOf? tokenIdOf: BigInt(0)
+        if(theTokenId == BigInt(0)) { return }
+        setTokenId(BigInt(parseInt(theTokenId.toString())))
+        setNewTokenImageURI(ipfsGetRet.image)
+      }
+      getNewToken()
+    }
+  }, [newTokenJSON, finalTokenUri, ipfsGet, ipfsGetIsSubmitting, tokenIdOf])
```

Comment:
and same things here.
can you separate useEffect hooks to a different file?
and I don't think all of the dependency is necessary. 
what is the purpose of this useEffect ?

Reply:
I made two different hooks to have them run sequentially. Do you think I can `await` everything and put these two into a single hook (separate file)?